backends/cuda/ceed-cuda-operator.c

```diff
@@ -649,25 +668,515 @@ static int CeedOperatorLinearAssembleQFunction_Cuda(CeedOperator op,
+        numemodeout += dim;
+        break;
+      case CEED_EVAL_WEIGHT:
+      case CEED_EVAL_DIV:
+      case CEED_EVAL_CURL:
+        break; // Caught by QF Assembly
+      }
+    }
+  }
+
+  // Operator data struct
+  CeedOperator_Cuda *impl;
+  ierr = CeedOperatorGetData(op, (void *)&impl); CeedChk(ierr);
+  ierr = CeedCalloc(1, &impl->diag); CeedChk(ierr);
+  CeedOperatorDiag_Cuda *diag = impl->diag;
+  diag->basisin = basisin;
+  diag->basisout = basisout;
+  diag->emodein_h = emodein;
+  diag->emodeout_h = emodeout;
+  diag->numemodein = numemodein;
+  diag->numemodeout = numemodeout;
+
+  // Assemble kernel
+  CeedInt nnodes, nqpts;
+  ierr = CeedBasisGetNumNodes(basisin, &nnodes); CeedChk(ierr);
+  ierr = CeedBasisGetNumQuadraturePoints(basisin, &nqpts); CeedChk(ierr);
+  diag->nnodes = nnodes;
+  ierr = CeedCompileCuda(ceed, diagonalkernels, &diag->module, 5,
+                         "NUMEMODEIN", numemodein,
+                         "NUMEMODEOUT", numemodeout,
+                         "NNODES", nnodes,
+                         "NQPTS", nqpts,
+                         "NCOMP", ncomp
+                        ); CeedChk_Cu(ceed, ierr);
+  ierr = CeedGetKernelCuda(ceed, diag->module, "linearDiagonal",
+                           &diag->linearDiagonal); CeedChk_Cu(ceed, ierr);
+  ierr = CeedGetKernelCuda(ceed, diag->module, "linearPointBlockDiagonal",
+                           &diag->linearPointBlock);
+  CeedChk_Cu(ceed, ierr);
+
+  // Basis matrices
+  const CeedInt qBytes = nqpts * sizeof(CeedScalar);
+  const CeedInt iBytes = qBytes * nnodes;
+  const CeedInt gBytes = qBytes * nnodes * dim;
+  const CeedInt eBytes = sizeof(CeedEvalMode);
+  const CeedScalar *interpin, *interpout, *gradin, *gradout;
+
+  // CEED_EVAL_NONE
+  CeedScalar *identity = NULL;
+  bool evalNone = false;
+  for (CeedInt i=0; i<numemodein; i++)
+    evalNone = evalNone || (emodein[i] == CEED_EVAL_NONE);
+  for (CeedInt i=0; i<numemodeout; i++)
+    evalNone = evalNone || (emodeout[i] == CEED_EVAL_NONE);
+  if (evalNone) {
+    ierr = CeedCalloc(nqpts*nnodes, &identity); CeedChk(ierr);
+    for (CeedInt i=0; i<(nnodes<nqpts?nnodes:nqpts); i++)
+      identity[i*nnodes+i] = 1.0;
+    ierr = cudaMalloc((void **)&diag->identity, iBytes); CeedChk_Cu(ceed, ierr);
+    ierr = cudaMemcpy(diag->identity, identity, iBytes,
+                      cudaMemcpyHostToDevice); CeedChk_Cu(ceed, ierr);
+  }
+
+  // CEED_EVAL_INTERP
+  ierr = CeedBasisGetInterp(basisin, &interpin); CeedChk(ierr);
+  ierr = cudaMalloc((void **)&diag->interpin, iBytes); CeedChk_Cu(ceed, ierr);
+  ierr = cudaMemcpy(diag->interpin, interpin, iBytes,
+                    cudaMemcpyHostToDevice); CeedChk_Cu(ceed, ierr);
+  ierr = CeedBasisGetInterp(basisout, &interpout); CeedChk(ierr);
+  ierr = cudaMalloc((void **)&diag->interpout, iBytes); CeedChk_Cu(ceed, ierr);
+  ierr = cudaMemcpy(diag->interpout, interpout, iBytes,
+                    cudaMemcpyHostToDevice); CeedChk_Cu(ceed, ierr);
+
```

Comment:
Following the usual cuda backend notation, I think the `identity`, `interpin/out` and `gradin/out` members of `diag` should be named `d_indentity`, `d_interpin`, etc. to indicate that they are on the device.

Reply:
Agreed and updated. Thank you.